sanic/headers.py

```diff
@@ -24,14 +24,132 @@ def parse_content_header(value: str) -> typing.Tuple[str, Options]:
+
+
+# https://tools.ietf.org/html/rfc7230#section-3.2.6 and
+# https://tools.ietf.org/html/rfc7239#section-4
+# This regex is for *reversed* strings because that works much faster for
+# right-to-left matching than the other way around. Be wary that all things are
+# a bit backwards! _rparam matches forwarded pairs alike ";key=value"
+_rparam = re.compile(f"(?:{_token}|{_quoted})={_token}\\s*($|[;,])", re.ASCII)
+
+
+def parse_forwarded(headers, config):
+    """Parse HTTP Forwarded headers.
+    Accepts only the last element with secret=`config.FORWARDED_SECRET`
+    :return: dict with matching keys (lowercase) and values, or None.
+    """
+    header = headers.getall("forwarded", None)
+    secret = config.FORWARDED_SECRET
+    if header is None or not secret:
+        return None
+    header = ",".join(header)  # Join multiple header lines
+    if secret not in header:
+        return None
+    # Loop over <separator><key>=<value> elements from right to left
+    ret = sep = pos = None
+    found = False
+    for m in _rparam.finditer(header[::-1]):
+        # Start of new element? (on parser skips and non-semicolon right sep)
+        if m.start() != pos or sep != ";":
+            if found:
+                return normalize(ret)
+            ret = {}
+        pos = m.end()
+        val_token, val_quoted, key, sep = m.groups()
+        key = key.lower()[::-1]
+        val = (val_token or val_quoted.replace('"\\', '"'))[::-1]
+        ret[key] = val
+        if key == "secret" and val == secret:
+            found = True
+        if found and sep != ";":
+            return normalize(ret)
+    return normalize(ret) if found else None
+
+
+def parse_xforwarded(headers, config):
+    """Parse X-Real-IP, X-Scheme and X-Forwarded-* headers."""
+    proxies_count = config.PROXIES_COUNT
+    if not proxies_count:
+        return None
+    h1, h2 = config.REAL_IP_HEADER, config.FORWARDED_FOR_HEADER
+    addr = h1 and headers.get(h1)
+    forwarded_for = h2 and headers.getall(h2, None)
+    if not addr and forwarded_for:
+        assert proxies_count == -1 or proxies_count > 0, config.PROXIES_COUNT
+        # Combine, split and filter multiple headers' entries
+        proxies = (p.strip() for h in forwarded_for for p in h.split(","))
+        proxies = [p for p in proxies if p]
+        try:
+            addr = proxies[-proxies_count] if proxies_count > 0 else proxies[0]
+        except IndexError:
+            return None
+    if not addr:
+        return None
+    other = (
+        (key, headers.get(header))
+        for key, header in (
+            ("proto", "x-scheme"),
+            ("proto", "x-forwarded-proto"),  # Overrides X-Scheme if present
+            ("host", "x-forwarded-host"),
+            ("port", "x-forwarded-port"),
+            ("path", "x-forwarded-path"),
+        )
+    )
+    return normalize({"for": addr, **{k: v for k, v in other if v}})
+
+
+_ipv6 = "(?:[0-9A-Fa-f]{0,4}:){2,7}[0-9A-Fa-f]{0,4}"
```

Comment:
We might need to make this regex a bit stronger if we are to rely on the regex.  (I am sure you have thought of this already) This would be a good to have. Otherwise, we might as well use the `ipaddress` module itself. 

xref: http://vernon.mauery.com/content/projects/linux/ipv6_regex

Reply:
I would use ipaddress but its implementation is awfully slow.

Reply:
@Tronic Really? I wasn't aware of that. Internally at work, I have a custom class called `IP`which does most of the things for me and it's based out of RegEx. But since `ipaddress` was part of python core I thought might as well try. But if it's really slow, then we can stick to regex.

Reply:
These are not entirely comparable as they do different things, but for reference:
```
%timeit ipaddress.ip_address("2606:4700:4700::1001")                   
8.76 µs ± 48.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit ipv6_re.match("2606:4700:4700::1001")                            
588 ns ± 1.34 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit parse_forwarded(headers, config)  # returns {'secret': 'mySecret', 'for': '[2606:4700:4700::1001]'}
5.75 µs ± 26.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Reply:
@Tronic That's interesting. Well, if we are using the RegEx, can we please have a bit more strict version of this RegEx? This regex can easily return false positives

Reply:
I don't see any practical way to make it more strict than it already is. Also, this is only meant to tell IPv4 and IPv6 addresses and hostnames apart, and as far as I can tell, there cannot be any false positives there. Neither IPv4 or hostnames can have 2-7 colons in them. The closest that I can think of is a hexadecimal-looking hostname in a host header like `cafe:8000` which still only has one colon, and IPv6 host should anyway be bracketed like `[f00d::cafe]:8000`.

Reply:
A somewhat bigger concern is that IPv6 are not normalized to compressed form. For this it might be justified to use ipaddress module even though running an address through it takes more time than parsing the entire headers.